---
title: "Differential splicing analysis using MAST"
author: "Mira Sohn"
output:
    html_document:
        code_folding: hide
        df_print: paged
        toc: true
        toc_float: true
        toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE,
                      message=FALSE,
                      cache.lazy=FALSE)
```

Last run: `r date()`

This workflow is designed to conduct differential splicing (DS) analysis
using [MAST](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5).

Refer to the following resources for technical details:

- [Documentation](https://rglab.github.io/MAST/index.html)
- [GitHub](https://github.com/RGLab/MAST/)

```{r libraries}
# Libraries
library(parallel)
library(tidyverse)
library(ggplot2)
library(reticulate)
library(parallel)
library(yaml)
library(reshape2)
library(limma)
library(NMF)
library(MAST)
library(RColorBrewer)
library(scater)
library(plotly)

# Additional options
set.seed(2570)
source('../config/helpers.R')
use_condaenv("../../../menv")

# Specify the number of available cores
n_cores <- future::availableCores()
options(mc.cores=n_cores)
```

```{r config}

# --------------------------------------------------------------------------------
# This chunk specifies input paths and variables in R
# --------------------------------------------------------------------------------

# Path to input matrix
mtx_path <- "../results/junction_counts/thalamus_sc_perind_numers.counts.gz"

# Path to metadata table
meta_path <- "../results/barcodes.tsv"

# Path to Snakemake config
snakemake_config <- "../config/config.yaml"

# Paths to AnnData
adata_paths <- read_yaml(snakemake_config)[['adata']] %>%
    map(~paste0("../", .x))

# Path to output directory
out_dir <- "ds"

# Create directories if missing
for (p in c(out_dir)) {
    if (! dir.exists(p)) { dir.create(p, recursive=TRUE) }
}

# Specify metadata columns for factors and factor levels
factor_groups <- list(
    general_disease=c('Control', 'FTD', 'AD'),
    cell_type=c('ExNeu1', 'ExNeu2'),
    study=c('Marsan', 'Biogen', 'Mathys'),
    study_specific_disease_specific=c('Control-Marsan',
                                      'Control-Biogen',
                                      'Control-Mathys',
                                      'FTD-Marsan',
                                      'FTD-Biogen',
                                      'AD-Mathys'),
    study_disease_specific_celltype=c('Control-Marsan_ExNeu1',
                                      'Control-Marsan_ExNeu2',
                                      'Control-Biogen_ExNeu1',
                                      'Control-Biogen_ExNeu2',
                                      'Control-Mathys_ExNeu1',
                                      'Control-Mathys_ExNeu2',
                                      'FTD-Marsan_ExNeu1',
                                      'FTD-Marsan_ExNeu2',
                                      'FTD-Biogen_ExNeu1',
                                      'FTD-Biogen_ExNeu2',
                                      'AD-Mathys_ExNeu1',
                                      'AD-Mathys_ExNeu2')
)

```

```{python ppackages}
import anndata as ad
import scanpy as sc
import pandas as pd
import numpy as np
import subprocess
import os
```


# Loading input data {.tabset}

The number of splicing junctions was counted based on cellranger-aligned reads mapped to the following
genes:

```{r genes_tested}
print(read_yaml(snakemake_config)[['genes']])
```

These counts were compiled into a junction-by-barcode matrix for differential testing.


```{python import_adata}

# --------------------------------------------------------------------------------
# This chunk imports input data and add the junction matrix to AnnData
# --------------------------------------------------------------------------------

# reticulate::repl_python()

# Import AnnData objects
adata_dic = {celltype: ad.read_h5ad(path) for celltype, path in r.adata_paths.items()}

# Import junction matrix
counts = pd.read_csv(r.mtx_path, sep=" ")

# Prep a function calculating junction length in Kb
def size_kb(coord):
    # Split the input coordinate. Assume the following format: 
    # "chr8:79611214:79616822:clu_1_+"
    parts = coord.split(":")
    start = int(parts[1])
    end = int(parts[2])
    # Calculate between the start and end sites in Kb
    distance = abs(end - start) / 1000
    return distance

for celltype in adata_dic.keys():
    # Retrieve an AnnData corresponding to the celltype
    ann = adata_dic[celltype]
    # Extract cell barcodes
    target_cells = list(ann.obs.index)
    # Slice the junction count matrix based on the extracted barcodes
    counts_subset = counts[target_cells]
    # Ensure that cell barcodes match between the count matrix and AnnData
    try: 
        # Add the raw junction count matrix to AnnData
        # NOTE: This requires barcode-by-junction as input
        ann.obsm['junctions_raw'] = counts_subset.T
        # Add log2-transformed junction matrix to AnnData
        ann.obsm['junctions_log2'] = np.log2(counts_subset.T + 1)
    except TypeError:
        print(f"Dimensions don't match between the two matrices for {celltype}!")

    # Calculate TPM
    # 1. Divide the read counts by the length of each gene in kilobases. This gives you reads 
    # per kilobase (RPK).
    # 2. Count up all the RPK values in a sample and divide this number by 1,000,000. This is 
    # your “per million” scaling factor.
    # 3. Divide the RPK values by the “per million” scaling factor. This gives you TPM.

    # Calculate the size of junctions
    length_list = [size_kb(c) for c in counts_subset.index]
    # Divide the raw counts by junction size, returning RPK
    counts_rpk = counts_subset.div(length_list, axis=0)
    # Calculate barcode-wise per million scaling factor
    per_barcode_million = counts_rpk.sum(axis=0) / 1_000_000
    # Convert into log2TPM scale with a pseudocount of 1
    tpm = counts_rpk.div(per_barcode_million, axis=1)
    log2tpm = np.log2(tpm + 1)
    # Replace missing values (NaN) with 1
    log2tpm = log2tpm.fillna(0)
    # Add log2TPM junctions to AnnData
    ann.obsm['junctions_log2TPM'] = log2tpm.T

    # Specify the path to updated AnnData
    updated_path = f"{r.out_dir}/{celltype}_JUNCTIONS.h5ad"
    # Save updated AnnData
    ann.write_h5ad(updated_path)

    # Replace the input AnnData with updated obj
    adata_dic[celltype] = ann

```

Input AnnData objects for each celltype:

```{r infile_info}

for (name in names(adata_paths)) {
    print(paste0(name, ": ", adata_paths[[name]]))
}

```

Refer to the following data added to the `AnnData` objects:

- `AnnData.obsm["junctions_raw"]`: raw junction counts
- `AnnData.obsm["junctions_log2"]`: log2-transformed raw junction counts with a pseudocount of 1
- `AnnData.obsm["junctions_log2TPM"]`: log2-transformed transcripts per million (TPM) 
  with a pseudocount of 1

Updated `AnnData` objects are saved as:

```{r outfile_info}

for (name in names(adata_paths)) {
    print(paste0(name, ": ", out_dir, "/", name, "_JUNCTIONS.h5ad"))
}

```

Input matrices are prepared as *log2-transformed TPM values (with a pseudocount of 1)* 
in the `SCA` object. The following table shows the number of barcodes for all cells (`all`) or
celltype-specific subsets (`ExNeu1` and `ExNeu2`). The `N_barcodes` column represents the number of 
barcodes in each object.

```{r prep_sca}

# --------------------------------------------------------------------------------
# This chunk builds an SCA obj. Refer to the following instructions:
# - https://rglab.github.io/MAST/reference/FromMatrix.html
# - https://rglab.github.io/MAST/articles/MAST-interoperability.html
# This chunk requires manual intervention to some extent.
# --------------------------------------------------------------------------------

# Extract metadata for all input cells
adata_meta <- map(py$adata_dic,
                  ~.x[['obs']] %>% rownames_to_column('barcode')) %>%
    bind_rows() %>%
    mutate(study_disease_specific_celltype=paste0(study_specific_disease_specific,
                                                  "_",
                                                  cell_type))

rownames(adata_meta) <- adata_meta[['barcode']]

# Refactor columns of interest
for (name in names(factor_groups)) {
    adata_meta[[name]] <- factor(adata_meta[[name]],
                                 levels=factor_groups[[name]])
}

# Ensure that the number of barcodes match between the matrix and metadata
if (nrow(adata_meta) != ncol(py$counts)) {
    stop("The number of barcodes is wrong!")
} else {
    # Reorder rows of the metadata 
    adata_meta <- adata_meta[colnames(py$counts), ]
}

# Prep feature table
feature_meta <- data.frame(junction=rownames(py$counts))

# Extract log2TPM-transformed counts from AnnData objects
l2t_mat <- map(py$adata_dic, ~.x$obsm[['junctions_log2TPM']]) %>%
    bind_rows() %>%
    t()

# Clean orders for row/column
l2t_mat <- l2t_mat[rownames(py$counts), colnames(py$counts)]

# Build an sca obj using log2TPM-transformed counts
sca <- FromMatrix(l2t_mat,
                  cData=adata_meta,
                  fData=feature_meta,
                  check_sanity=FALSE)

# Add raw counts to the sca obj
assays(sca)$counts <- py$counts

# Create a list storing objects for all and celltype-specific barcodes
sca_list <- list(all=sca)
for (cell_x in factor_groups[['cell_type']]) {
    barcode_x <- adata_meta[adata_meta[['cell_type']] == cell_x,] %>%
        rownames()
    sca_x <- sca[, barcode_x]
    sca_list[[cell_x]] <- sca_x
}

# Print the number of barcodes for each obj
df <- data.frame(cells=names(sca_list),
                 N_barcodes=sapply(sca_list, function(o) ncol(o))) %>%
    remove_rownames()

knitr::kable(df)

# How to access to raw or transformed counts
# - Transformed counts: `assay(sca)` or `assays(sca)$et`
# - Raw counts: `assays(sca)$counts`
```

# Quality Control (QC)

## Pre-filtering barcodes

Cells are removed if none of the junctions are detected. The number of remaining barcodes after filtering
is shown in the `N_remaining` column.

```{r prefiltering, results='asis'}

# --------------------------------------------------------------------------------
# This chunk removes barcodes with zero junction counts.
# --------------------------------------------------------------------------------

for (name in names(sca_list)) {

    # TRUE for non-zero and FALSE for zero-counted cells
    o <- sca_list[[name]]
    nonzero_barcodes <- assay(o) %>%
        apply(2, function(x) sum(x) > 0)

    # Remove zero-counted cells from the obj
    o_filt <- o[, nonzero_barcodes]

    # Print the number of cells after and after filtering
    df[df$cells == name, 'N_remaining'] <- ncol(o_filt)

    # Update the objects in the `sca_list`
    sca_list[[name]] <- o_filt
}

knitr::kable(df)
```

## Heatmap {.tabset}

Cell-to-cell similarity is explored using a heatmap generated from non-zero cells, with *Euclidean distance* 
represented by the color scale. Darker blue indicates less distance (i.e., greater similarity)
in splice junction profiling.

```{r qc_heatmap, results='asis', fig.height=12, fig.width=12, cache=TRUE}

# --------------------------------------------------------------------------------
# This chunk creates cell-to-cell similarity heatmap
# --------------------------------------------------------------------------------

for (name in names(sca_list)) {
    cat('###', name, '{.tabset}\n\n')
    # Retrieve obj
    o <- sca_list[[name]]
    # Retrieve log2TPM matrix
    sampleDistsMatrix <- as.matrix(dist(t(assay(o))))

    # Set color to be displayed
    colors <- colorRampPalette(rev(RColorBrewer::brewer.pal(9, 'Blues')))(255)

    # Prep metadata for heatmap
    heatmap_meta <- as.data.frame(colData(o)[, names(factor_groups)])

    # Print heatmap
    print(pheatmap::pheatmap(
        sampleDistsMatrix,
        scale='none',
        color=colors,
        annotation_col=heatmap_meta,
        show_rownames=FALSE,
        show_colnames=FALSE,
        cluster_cols=TRUE,
        border_color=NA
    )
    )
    cat('\n\n')
}
```

## UMAP {.tabset}

Junction profiling is explored across the cells on UMAP.

```{r prep_umap}

# --------------------------------------------------------------------------------
# This chunk prepares log2TPM count matrices and metadata for each subset
# --------------------------------------------------------------------------------

count_list <- lapply(sca_list,
                     function(o) list(mat=as.data.frame(assay(o)),
                                      meta=as.data.frame(colData(o))))
```

```{python umap}

# --------------------------------------------------------------------------------
# This chunk runs UMAP and prepares UMAP coordinates
# --------------------------------------------------------------------------------

umap_dic = {}

for subset in r.count_list.keys():
    # Retrieve input for AnnData
    obj_x = r.count_list[subset]
    mat_x = obj_x['mat']
    meta_x = obj_x['meta']

    # Create an AnnData obj
    adata_x = ad.AnnData(X=mat_x.T, obs=meta_x)

    # Run PCA
    sc.tl.pca(adata_x)

    # Find neighbors
    sc.pp.neighbors(adata_x)

    # Run UMAP
    sc.tl.umap(adata_x)

    # Retrieve UMAP coordinates (2D)
    umap_coord = adata_x.obsm['X_umap']

    umap_dic[subset] = umap_coord
```

```{r present_umap, results='asis', fig.height=6, fig.width=6}

# --------------------------------------------------------------------------------
# This chunk plots UMAP
# --------------------------------------------------------------------------------

# Consolidate metadata and UMAP coordinates in the same data frame
umap_list <- lapply(py$umap_dic,
                    function(df) as.data.frame(df) %>% set_names(c("UMAP1", "UMAP2"))) %>%
    map2(count_list, ~cbind(.y[['meta']], .x))

# Plot
for (subset in names(umap_list)) {
    cat('###', subset, '{.tabset}\n\n')
    for (name in names(factor_groups)) {
        cat('####', name, '\n\n')
        df <- umap_list[[subset]][, c(name, "UMAP1", "UMAP2", "barcode")]
        p <- ggplot(df, aes_string(x='UMAP1', y='UMAP2', color=name)) +
            geom_point(alpha=0.7) +
            theme_bw()
        print(p)
        cat('\n\n')
    }
}
```


# Session info

```{r session_info, collapse=FALSE}
sessionInfo()
```



