import os
import sys
import subprocess
import re
import numpy as np
import pandas as pd
import pysam
import anndata as ad

#################################### USER CONFIGURATION ####################################

# Path to config file
configfile: 'config/config.yaml'

# Create the output directory
outdir = config['outdir']
if not os.path.exists(outdir):
    os.makedirs(outdir)

# Path to barcode table
barcode_file = f"{outdir}/barcodes.tsv"

#################################### PREP BARCODES AND SAMPLETABLE ####################################

# NOTE: 
# - Manually prep data frames for metadata and sampletable
# - Ensure that the meta_df includes the three required columns
#   specified in config['key_col'], config['index_col'], and config['bam_col']

# Prep sampletable
sampletable = pd.read_csv(config['sampletable']).set_index(config['index_col'], drop=False)
sampletable[config['bam_col']] = config['bam_dir'] + "/" + sampletable.index + '_possorted_genome_bam.bam'

# Break if input bams don't align with sampletable
missing_bams = [b for b in sampletable[config['bam_col']] if not os.path.isfile(b)]
if len(missing_bams) > 0:
    raise ValueError(f"{missing_bams} not found in the directory!")

# Prep barcode table
meta_list = []
for celltype, path in config['adata'].items():
    # Read AnnData
    meta_df = ad.read_h5ad(path).obs[list(config['all_cols'])]
    # Add new columns for celltypes and barcodes
    meta_df['celltype'] = celltype
    meta_df['barcode'] = meta_df.index
    # Remove sample suffixes from the barcodes
    barcode_df = meta_df['barcode'].str.split('_', expand=True)
    barcode_df.columns = ['barcode', config['index_col']]
    # Add new columns to the metadata
    meta_df['barcode_sc'] = meta_df['barcode']
    meta_df['barcode'] = barcode_df['barcode']
    meta_df[config['index_col']] = barcode_df[config['index_col']]
    meta_list.append(meta_df)

# Save all barcodes in a tsv file
meta_df = pd.concat(meta_list, axis=0)

# print(meta_df.head(3))
# status   study Genetics  AgeDeath  Disease BrainRegion Sex general_disease study_disease study_specific_disease_specific celltype             barcode                   barcode_sc  SampleID
# CGTCTACTCAAAGTAG-1_Control1  Control  Marsan  Control        68  Control    Thalamus   F         Control       Control                  Control-Marsan   ExNeu1  CGTCTACTCAAAGTAG-1  CGTCTACTCAAAGTAG-1_Control1  Control1
# CCATGTCAGTTGTAGA-1_Control1  Control  Marsan  Control        68  Control    Thalamus   F         Control       Control                  Control-Marsan   ExNeu1  CCATGTCAGTTGTAGA-1  CCATGTCAGTTGTAGA-1_Control1  Control1
# CAGGTGCCAGGTTTCA-1_Control1  Control  Marsan  Control        68  Control    Thalamus   F         Control       Control                  Control-Marsan   ExNeu1  CAGGTGCCAGGTTTCA-1  CAGGTGCCAGGTTTCA-1_Control1  Control1

# print(len(meta_df))
# 27719
#
## Add bam file paths to the metadata table and save as a tsv file
meta_df = pd.merge(sampletable.loc[:, [config['index_col'], config['key_col'], config['bam_col']]].reset_index(drop=True),
                   meta_df,
                   how="right",
                   on=[config['index_col'], config['key_col']])
meta_df.to_csv(barcode_file, sep="\t", index=False)

# print(meta_df.head(3))
# SampleID        study_specific_disease_specific bam     status  study   Genetics        AgeDeath        Disease BrainRegion     Sex     general_disease study_disease   celltype        barcode barcode_sc
# Control1        Control-Marsan  ../../input/thalamus_excitatory/bam/Control1_possorted_genome_bam.bam   Control Marsan  Control 68      Control Thalamus        F       Control Control ExNeu1  CGTCTACTCAAAGTAG-1      CGTCTACTCAAAGTAG-1_Control1
# Control1        Control-Marsan  ../../input/thalamus_excitatory/bam/Control1_possorted_genome_bam.bam   Control Marsan  Control 68      Control Thalamus        F       Control Control ExNeu1  CCATGTCAGTTGTAGA-1      CCATGTCAGTTGTAGA-1_Control1

##################################### PREP FUNCTIONS ####################################

# Function to extract SQ-header from a bam file
def extract_headers_SQ(bam):
    # Collect unique SQ header reads in a set
    headers = set()
    with pysam.AlignmentFile(bam, "rb") as f:
        for line in f.header['SQ']:
            headers.add((line['SN'], line['LN']))
    return headers

# Function to select input bam files required for output 
def select_bam(wildcards):
    filtered_df = meta_df[(meta_df['barcode_sc'] == wildcards.sc_barcode)]
    samplename = filtered_df['SampleID'].unique()
    return "{}/{}_possorted_genome_bam.bam".format(config['bam_dir'], samplename)


#################################### RULES ####################################

rule all:
    input:
        expand("{outdir}/sam/{sc_barcode}.sam",
               outdir=config['outdir'],
               sc_barcode=list(set(meta_df['barcode_sc'])))

rule create_header:
    """
    This rule checks whether all headers (@SQ) are identical across the samples
    and creates header
    """
    input:
        bams = lambda wildcards: [b for b in set(meta_df['bam']) if pd.notnull(b) and os.path.isfile(b)]
    output:
        sam = "{outdir}/header.sam"
    resources:
        mem_mb = 1024 * 20,
        disk_mb = 1024 * 10,
        runtime = 30
    run:
        header_list = []
        for bam in input.bams:
            header_set = extract_headers_SQ(bam)
            if header_set not in header_list:
                header_list.append(header_set)
        if len(header_list) != 1:
            raise ValueError("Headers are discordant across the samples!")
        else:
            shell('samtools view -H {input.bams[0]} | grep "^@SQ" > {output.sam}')

rule prep_sam:
    """
    This rule preps barcodes per barcode"
    """
    input:
        header = "{outdir}/header.sam",
        bam = select_bam
    output:
        sam = "{outdir}/sam/{sc_barcode}.sam"
    resources:
        mem_mb = 1024 * 20,
        disk_mb = 1024 * 10,
        runtime = 60
    threads: 8
    params:
        genes = config['genes']
    run:
        # Sanity check for the input bam file
        if not os.path.exists(input.bam):
            raise ValueError("Input bam file not found!")
        # Extract a cellranger barcode based on the sample-associated barcode
        df = meta_df[(meta_df[config['barcode_sc']] == wildcards.sc_barcode)]
        target_barcode = df.barcode
        # Prep a list for genes of interest
        gene_list = [f"GN:Z:{g.strip()}" for g in params.genes]
        # Join genes into a single string (w header)
        gene_joined = "|".join(gene_list)
        # Specify a temporary sam file 
        temp_sam = f"{wildcards.outdir}/{wildcards.sc_barcode}_temp.sam"
        # Run samtools with or without module load samtools/1.19
        shell(
            """
            cat {input.header} > {output.sam}
            samtools view -@ {threads} {input.bam} | grep "{target_barcode}" > {temp_sam} || true
            cat {temp_sam} | grep -E "{gene_joined}" >> {output.sam} || true
            rm -f {temp_sam}
            """
            )
